src/main/java/com/datapipe/jenkins/vault/credentials/common/VaultHelper.java

```diff
@@ -23,76 +25,111 @@
 import javax.annotation.Nonnull;
 import jenkins.model.GlobalConfiguration;
 import jenkins.model.Jenkins;
+import jenkins.security.SlaveToMasterCallable;
 import org.apache.commons.lang.StringUtils;
 
+import static org.apache.commons.lang.StringUtils.isNotEmpty;
+
 public class VaultHelper {
+
     private static final Logger LOGGER = Logger.getLogger(VaultHelper.class.getName());
 
     static String getVaultSecret(@Nonnull String secretPath, @Nonnull String secretKey, @Nonnull Integer engineVersion) {
-        Jenkins jenkins = Jenkins.getInstanceOrNull();
-        if (jenkins == null) {
-            LOGGER.warning("Cannot retrieve secret becuase Jenkins.instance is not available");
-            return null;
+        try {
+            String secret;
+            SecretRetrieve retrieve = new SecretRetrieve(secretPath, secretKey, engineVersion);
+
+            Channel channel = Channel.current();
+            if (channel == null) {
+                secret = retrieve.call();
+            } else {
+                secret = channel.call(retrieve);
```

Comment:
This may have side effects for instance we have limited our app secret to only allow Jenkins master based on IP scope making these calls.

Suddenly having agents making the request would mean we would have to change our app secret access.
I assume others have a similar setup.

Reply:
I'm not sure to fully understand your concern.

Can you help me to reproduce this setup?

Also, what kind of changes would it imply in the code? Do you have code samples to point me at?

Thanks.

Reply:
I am especially talking about https://www.vaultproject.io/api-docs/auth/approle/#inlinecode-secret_id_bound_cidrs

Reply:
Ok, I see.

I'm afraid the plugin doesn't behave as you would expect in this particular setup, even with the current version of the plugin. Because most of the secret retrievals are done on the master, not the agent.

I struggled with this a couple of hours while implementing the *Vault SSH username with private key* credential.
I had a multi-branch pipeline, with a Git source using a *Vault SSH username with private key* credential (same behavior with *Vault username-password* credential). Everything worked fine in a master only configuration, but it broke as soon as I tested it in a master / agent configuration.
The *branch scan* part of the pipeline was successful but the *clone* part failed. In the system logs, the plugin warned *Cannot retrieve secret becuase Jenkins.instance is not available*, from [VaultHelper, line 34](https://github.com/jenkinsci/hashicorp-vault-plugin/blob/60064f031472a8fdb7e453633a82ada3e03b9dfc/src/main/java/com/datapipe/jenkins/vault/credentials/common/VaultHelper.java#L34).
It turns out that the *branch scan* is executed on the master, while the *clone* part is executed on the agent. Apparently, the `withCredentials()` step also runs on the master.
After digging, I found out that `Jenkins.instance` was `null` on the agent. That's why the secret retrieval didn't work and returned `null`.

To sum up, you can't authorize the access to a Vault secret based on the IP address, because the secrets are retrieved by the master.
It might be an issue, but it deserves a separate discussion.